pkg/controller/trainingruntime_controller.go

```diff
@@ -0,0 +1,149 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"iter"
+	"slices"
+	"time"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	trainer "github.com/kubeflow/trainer/pkg/apis/trainer/v1alpha1"
+	"github.com/kubeflow/trainer/pkg/constants"
+	idxer "github.com/kubeflow/trainer/pkg/runtime/indexer"
+	"github.com/kubeflow/trainer/pkg/trainjob"
+)
+
+const (
+	updateChBuffer = 10
+)
+
+type TrainingRuntimeReconciler struct {
+	log                      logr.Logger
+	client                   client.Client
+	recorder                 record.EventRecorder
+	nonRuntimeObjectUpdateCh chan event.TypedGenericEvent[iter.Seq[types.NamespacedName]]
+}
+
+var _ reconcile.Reconciler = (*TrainingRuntimeReconciler)(nil)
+var _ TrainJobWatcher = (*TrainingRuntimeReconciler)(nil)
+
+func NewTrainingRuntimeReconciler(cli client.Client, recorder record.EventRecorder) *TrainingRuntimeReconciler {
+	return &TrainingRuntimeReconciler{
+		log:                      ctrl.Log.WithName("trainingruntime-controller"),
+		client:                   cli,
+		recorder:                 recorder,
+		nonRuntimeObjectUpdateCh: make(chan event.TypedGenericEvent[iter.Seq[types.NamespacedName]], updateChBuffer),
+	}
+}
+
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes,verbs=create;get;list;watch;update;patch;delete
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=trainer.kubeflow.org,resources=trainingruntimes/finalizers,verbs=get;update;patch
+
+func (r *TrainingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var runtime trainer.TrainingRuntime
+	if err := r.client.Get(ctx, req.NamespacedName, &runtime); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	log := ctrl.LoggerFrom(ctx).WithValues("trainingRuntime", klog.KObj(&runtime))
+	ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling TrainingRuntime")
+
+	trainJobs := &trainer.TrainJobList{}
+	if err := r.client.List(ctx, trainJobs, client.InNamespace(runtime.Namespace), client.MatchingFields{
+		idxer.TrainJobRuntimeRefKey: runtime.Name,
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
So even if TrainJob is complete, we are waiting until all Jobs will be deleted before TrainingRuntime can be updated.
I am wondering how do we view the upgrade path for our users ?
For example, if user wants to update container image in the Runtime.

Reply:
Currently, we do not consider edge case. We just prevent the removal of TrainingRuntime to prevent running TrainJob disruption.
If we want to allow them to delete when all referenced TrainJob is finalized, we need to handle terminal states.
However, I'm not sure if I can finalize it within this week.

Reply:
Ok, so for now we require users to delete all of the TrainJobs if they want to upgrade TrainingRuntime, right ?

Reply:
Yes.
After https://github.com/kubeflow/trainer/pull/2619, runtimes are immutable.

Reply:
As we discussed previously, we need to introduce revision mechanism to support runtime updates.